autoload/neomake/makers/ft/sh.vim

```diff
@@ -4,14 +4,35 @@ function! neomake#makers#ft#sh#EnabledMakers() abort
     return ['sh', 'shellcheck']
 endfunction
 
-" IDEA: could be detected once from "shellcheck --help" (since older versions
-" support zsh, and newer might do so again).
-let s:shellcheck_supported = ['sh', 'bash', 'dash', 'ksh']
+function! s:shellcheck_determine_supported() abort
+    let s:shellcheck_supported = ['sh', 'bash', 'dash', 'ksh']
+
+    " zsh support was available from the first release of shellcheck and
+    " removed in version 0.3.6. Before shellcheck version 0.3.1 there was no
+    " way to get the version information.
+    let l:version_line = matchstr(split(system('shellcheck --version'), '\n'),
+                \ '^version:')
+    if v:shell_error != 0
+        let s:shellcheck_supported += ['zsh']
+    else
+        let l:version = matchstr(l:version_line, '^version:\s*\zs.*$')
+        if neomake#utils#CompareSemanticVersions(l:version, '0.3.6') == -1
+            let s:shellcheck_supported += ['zsh']
+        endif
+    endif
+endfunction
+call s:shellcheck_determine_supported()
```

Comment:
This should be done lazily, only when GetEnabledMakers is really called - and just once (if not cached already).